PF-2173 Add more filtering support

nmalfroy opened the pull request:
This PR introduces some new filtering concepts:
- Adds supports for `IN` operations
- Allows filtering on a list of flights
- Adds the ability to specify arbitrary boolean filters when filtering for flights.  This allows users to construct their own complex filters on input parameters.  E.g.
<pre><code><del>FlightFilter flightFilter =
        new FlightFilter()
            .setFilterInputBooleanOperationParameter(
                f ->
                    createOr(
                        createPredicate(
                            f.makeInputPredicate("foo", FlightFilterOp.EQUAL, null)),
                        createAnd(
                            f.makeInputPredicate("foo", FlightFilterOp.EQUAL, "val"),
                            f.makeInputPredicate("bar", FlightFilterOp.LESS_EQUAL, 42))));
</del>
FlightFilter filter = new FlightFilter(
    makeAnd(
        makePredicateFlightStatus(EQUAL, FlightStatus.ERROR),
        makePredicateInput("other_num", LESS_THAN, 456)),
        makeOr(
            makePredicateInput("email", EQUAL, "ddtest@gmail.com"),
            makePredicateInput("num", GREATER_THAN, 123)))
    .submittedTimeSortDirection(FlightFilterSortDirection.DESC);
</code></pre>
(note the tests have some good examples...happy to add more if needed)
- Improves support for `IS NULL` operations on input parameters (the parameter no longer needs to exist)
- Note: this gets rid of the form1-form2-form3 sql generation in favor or a single approach that chains together `EXISTS` statements.  The PR creates an index on the flight input key field to speed up the approach. I tested on a stairway DB with ~5M flights and ~40M rows in the flight inputs table and the performance is pretty good (~1.5 seconds when doing basic filtering)
- It's worth noting: I left the current mode of adding input filters too for backwards compatibility

nmalfroy committed:
Add support for querying with an IN clause

- Works for both flight fields and input fields

nmalfroy committed:
Add support for boolean expressions in filters

pshapiro4broad commented on stairway/src/main/java/bio/terra/stairway/impl/FlightFilterAccess.java:
```diff
@@ -226,17 +200,19 @@ private void makeFlightFilter(StringBuilder sb, String prefix) {
-      inter = " OR ";
+  // -- boolean expression methods
+  private String makeBooleanExpressionsFilters(FlightBooleanOperationExpression expression) {
+    switch (expression.getOperation()) {
+      case PREDICATE:
+        return makeInputPredicateSql(expression.getBasePredicate());
+      case AND, OR:
+        return "("
+            + expression.getExpressions().stream()
+                .map(this::makeBooleanExpressionsFilters)
+                .collect(Collectors.joining(expression.getOperation().getSql()))
+            + ")";
+      default:
+        throw new FlightFilterException("Unrecognized boolean operation");
     }
```
Use a switch expression here to avoid the need to `throw`.
```suggestion
    return switch (expression.getOperation()) {
      case PREDICATE -> makeInputPredicateSql(expression.getBasePredicate());
      case AND, OR -> "("
            + expression.getExpressions().stream()
                .map(this::makeBooleanExpressionsFilters)
                .collect(Collectors.joining(expression.getOperation().getSql()))
            + ")";
    }
```

nmalfroy replied:
This isn't on java 17 yet so that syntax doesn't work sadly

pshapiro4broad replied:
Oh wait, is stairway not on JDK 17 yet?

nmalfroy replied:
Yeah.  Libraries get converted later since you don't want to break dependencies and I don't think that all services that use it are on 17 yet

pshapiro4broad replied:
Actually stairway _is_ on JDK 17 as of https://github.com/DataBiosphere/stairway/pull/98

nmalfroy replied:
Oh weird...it was giving me sass

pshapiro4broad commented on stairway/src/main/java/bio/terra/stairway/impl/FlightFilterAccess.java:
```diff
@@ -226,17 +200,19 @@ private void makeFlightFilter(StringBuilder sb, String prefix) {
+      case PREDICATE:
+        return makeInputPredicateSql(expression.getBasePredicate());
+      case AND, OR:
+        return "("
+            + expression.getExpressions().stream()
+                .map(this::makeBooleanExpressionsFilters)
+                .collect(Collectors.joining(expression.getOperation().getSql()))
+            + ")";
```
You can simplify this by using the `joining()` form that supports prefix and suffix.
```suggestion
        return expression.getExpressions().stream()
                .map(this::makeBooleanExpressionsFilters)
                .collect(Collectors.joining(expression.getOperation().getSql(), "(", ")"));
```

nmalfroy replied:
Oh, interesting...I didn't realize you could prefix and suffix there too.  cool

pshapiro4broad commented on stairway/src/main/java/bio/terra/stairway/impl/FlightFilterAccess.java:
```diff
@@ -289,19 +278,66 @@ private void storeFlightPredicateValue(
-    return "(I.key = '"
+    String valuePredicate = "";
+    if (predicate.getDatatype() == Datatype.NULL) {
+      return "(EXISTS (SELECT 0 FROM flightinput I WHERE F.flightid = I.flightid AND I.key = "
```
If the subselect is the same, is there a way to refactor the string to avoid the code duplication?

nmalfroy replied:
It's not totally the same and pulling out the similar pieces I thought would make the code less readable

pshapiro4broad commented on stairway/src/main/java/bio/terra/stairway/impl/FlightFilterAccess.java:
```diff
@@ -289,19 +278,66 @@ private void storeFlightPredicateValue(
+            StairwayMapper.getObjectMapper()
+                .writeValueAsString(StairwayMapper.getObjectMapper().writeValueAsString(value)));
+      }
+      jsonValue = "[" + StringUtils.join(values, ",") + "]";
```
`StringUtils` is unnecessary here
```suggestion
      jsonValue = "[" + String.join(",", values) + "]";
```

pshapiro4broad commented on stairway/src/test/java/bio/terra/stairway/impl/EnumerateFlightsTest.java:
```diff
@@ -174,6 +176,82 @@ public void enumTest() throws Exception {
+        new FlightFilter()
+            .addFilterInputParameter("in2", FlightFilterOp.IN, Arrays.asList(pojo1, pojo2));
+    flightList = flightDao.getFlights(0, 100, filter);
+    checkResults("case 21", flightList, Arrays.asList("1", "2", "3", "4"));
```
It looks like this code supports using boolean operations and other operations (like `addFilterInputParameter()`)  in the same filter. If so, is this functionality tested? I didn't see a test for it here.

nmalfroy replied:
Good eye.  I'll add a test for that

pshapiro4broad commented on stairway/src/main/java/bio/terra/stairway/FlightFilter.java:
```diff
@@ -177,6 +213,35 @@ public FlightFilter submittedTimeSortDirection(
+
+    if (op == FlightFilterOp.IN) {
+      return new FlightFilterPredicate(key, op, value, Datatype.LIST, makeParameterName());
+    } else if (value == null) {
```
I'm wondering why you chose to do this, rather than an explicit op like `FlightFilterOp.IS_NULL`. Also, is there a way to check `NOT NULL`?

nmalfroy replied:
There was already logic for NULL datatypes so it seemed more appropriate to keep that

pshapiro4broad commented on stairway/src/main/java/bio/terra/stairway/FlightFilter.java:
```diff
@@ -233,7 +298,81 @@ public String getParameterName() {
+    }
+
+    public enum Operation {
+      PREDICATE(null),
```
It seems like you could support `NOT` by adding `PREDICATE_NOT` here.. not sure how hard it would be to do though, or how useful it would be.

nmalfroy replied:
100% I was debating doing it but the PR was getting kinda big already so I held off.  If folks think it would be useful, I'm happy to add it too (it's pretty easy)

nmalfroy committed:
Address Inital PR feedback

- More testing
- Code cleanup

ddietterich reviewed (changes requested):
I don't think the is the right filter interface. It would be better if there were one class hierarchy of boolean ops and field predicates that could be constructed by the caller. I want you to be able to say:
```
FlightFilter flightFilter =
  new FlightFilter()
    .makeOr(
       makeInputPredicate("foo", FlightFilterOp.EQUAL, null))
    .makeAnd(
       makeInputPredicate("foo", FlightFilterOp.EQUAL, "val"),
       makeInputPredicate("bar", FlightFilterOp.LESS_EQUAL, 42)));
```
I suggest making a varargs version of the constructor, so that we could write the expression without the dots (e.g., `.makeOr`). So the current constructor would stay as is and take the flat fluent `.make` form and the new constructor would be something like:
```
public FlightFilter(PredicateBaseClass... predicates)
```
and take allow constructing the predicate tree.

nmalfroy commented:
Yeah, the problem comes with nesting (like how do you nest an or clause within an and clause?  Which is a thing that TDR needs).  We basically need a thing that encompassed the current input predicates and the boolean operator

nmalfroy commented:
I don't think you could express the filter in the PR description with what you're proposing

ddietterich commented:
You have to have a base class that can be either a filter or a boolean. Then the (...) lists can take either, like this:
```
FlightFilter flightFilter =
  new FlightFilter(
     makeAnd(
       makeInputPredicate("foo", FlightFilterOp.EQUAL, null)),
       makeInputPredicate("fribble", FlightFilterOp.EQUAL, "Friendly"),
       makeOr(
         makeInputPredicate("foo", FlightFilterOp.EQUAL, "val"),
         makeInputPredicate("bar", FlightFilterOp.LESS_EQUAL, 42),
         makeAnd(
            makeInputPredicate("blah", FlightFilterOp.EQUAL, "Unfriendly"),
            makeInputPredicate("haha", FlightFilterOp.NOT_EQUAL, "funny"))));
```

nmalfroy commented:
I suppose we could so something like a nested builder model where there's a builder context.  So it could look something like:
```
FlightFilter flightFilter =
  new FlightFilter()
    .startOr()
       .addInputPredicate("foo", FlightFilterOp.EQUAL, null))
       .startAnd()
           .addInputPredicate("foo", FlightFilterOp.EQUAL, "val"),
           .addInputPredicate("bar", FlightFilterOp.LESS_EQUAL, 42)));
       .endAnd()
    .endOr()
```

That seems a little more unwieldy to me but it could work. What do you think?

ddietterich commented:
You don't like holding the list in an alternate FlightFilter constructor for some reason?

nmalfroy commented:
> You have to have a base class that can be either a filter or a boolean. Then the (...) lists can take either, like this:
> 
> ```
> FlightFilter flightFilter =
>   new FlightFilter(
>      makeAnd(
>        makeInputPredicate("foo", FlightFilterOp.EQUAL, null)),
>        makeInputPredicate("fribble", FlightFilterOp.EQUAL, "Friendly"),
>        makeOr(
>          makeInputPredicate("foo", FlightFilterOp.EQUAL, "val"),
>          makeInputPredicate("bar", FlightFilterOp.LESS_EQUAL, 42),
>          makeAnd(
>             makeInputPredicate("blah", FlightFilterOp.EQUAL, "Unfriendly"),
>             makeInputPredicate("haha", FlightFilterOp.NOT_EQUAL, "funny"))));
> ```

I can certainly do that but that would be a breaking change and I was trying to avoid that

nmalfroy commented:
A

> You don't like holding the list in an alternate FlightFilter constructor for some reason?

Oh, I don't mind that, I just was worried having a new base class which would be a breaking change. If we're ok with that, I can totally make a base class that encompasses the boolean operation and the input filter and then allow nesting that way

ddietterich commented:
Why would it be a breaking change?

I am imagining that the old constructor creates a top-level AND predicate and accepts the old style `.make` methods. The new style can set some state that would disallow the old style methods and only take the nested form.

nmalfroy commented:
Ah that's an interesting point!  Ok, lemme play with this and I'll update the PR.  Thanks for the suggestion!  I think this'll look a lot cleaner

nmalfroy commented:
Sorry, getting back to this today.  I'm actually not sure that this is going to work.  In the example that you've for makeAnd, makeOr, makePredicate are all being called from a static context (at least that's what it looks like).  The problem is that an input predicate needs to call the makeParameter method which is tied to an FlightFilter instance...not really sure how to work around short of using a provider pattern but then we're kind right back where we started

nmalfroy commented:
Actually, I think I got a thing working with a builder pattern.  I'll post an update in a few

nmalfroy committed:
Use a builder pattern and a common interface to specify boolean expressions in filters

nmalfroy commented:
Ok, I updated the pattern.  @ddietterich, curious what you think now?

nmalfroy commented:
@ddietterich Just sending a ping to see if you've had a chance to re-review

ddietterich commented on stairway/src/main/java/bio/terra/stairway/FlightFilter.java:
```diff
@@ -148,15 +193,27 @@ public FlightFilter addFilterFlightStatus(FlightFilterOp op, FlightStatus status
+   * @return {@code this}, for fluent style
+   * @throws FlightFilterException if predicate is not supplied
+   */
+  public FlightFilter setFilterInputBooleanOperationParameter(
```
InterlliJ thinks this call is unused. Delete?

nmalfroy replied:
yep. removed

ddietterich commented on stairway/src/main/java/bio/terra/stairway/FlightFilter.java:
```diff
@@ -23,10 +24,23 @@
+ *     makeAnd(
+ *         makePredicate("other_num", LESS_THAN, 456)),
+ *         makeOr(
+ *             makePredicate("email", EQUAL, "ddtest@gmail.com"),
```
The method is actually, `makeInputPredicate`. That raises the question, how do flight predicates fit into this?
I was expecting to see a `makeFlightPredicate` method to match the `makeInputPredicate` method that would integrate the flight predicates into the general boolean expression.

nmalfroy replied:
Yep, totally agreed.  I updated so that now predicates can be flight and input predicates

nmalfroy replied:
(I updated the doc string to show that too)

ddietterich commented on stairway/src/main/java/bio/terra/stairway/FlightFilter.java:
```diff
@@ -177,12 +234,76 @@ public FlightFilter submittedTimeSortDirection(
   }
 
-  public static class FlightFilterPredicate {
+  public interface FlightFilterPredicateInterface {
```
IIUC the only reason for the Builder model is to do name assignments. Seems like we could do that during filter SQL generation instead. It just happened to be easier to do it in the FlightFilter class using the original technique. It makes this code rather clunky.

pshapiro4broad replied:
I find the current code's use of `builder` to be confusing and doesn't seem to be doing much. The code has general support for  creating a parameter name but in the code base the only implementation of this is `FlightFilter.makeParameterName()`. Could the code instead call this method at the time the predicate is created?

pshapiro4broad replied:
I've refactored the code locally to no longer use `builder` and it seems to work fine (that is, all the stairway tests are still passing). The only downside I can see is that creating a filter like this is slightly more awkward. Instead of
```
    FlightFilter filter =
        new FlightFilter(
            makeOr(
                makePredicateFlightClass(FlightFilterOp.EQUAL, TestFlight.class),
                makePredicateInput("email", FlightFilterOp.EQUAL, "ddtest@gmail.com"),
                makeAnd(
                    makePredicateInput("name", FlightFilterOp.EQUAL, "dd"),
                    makePredicateInput("resource", FlightFilterOp.EQUAL, "resourceId"))));
```

I have
```
    FlightFilter filter = new FlightFilter();
    filter.setInputBooleanOperationExpression(
        filter.makeOr(
            filter.makePredicateFlightClass(FlightFilterOp.EQUAL, TestFlight.class),
            filter.makePredicateInput("email", FlightFilterOp.EQUAL, "ddtest@gmail.com"),
            filter.makeAnd(
                filter.makePredicateInput("name", FlightFilterOp.EQUAL, "dd"),
                filter.makePredicateInput("resource", FlightFilterOp.EQUAL, "resourceId"))));
```

pshapiro4broad replied:
If instead the parameter was generated during sql generation as @ddietterich suggests then the `make` methods could become static again.

pshapiro4broad replied:
I have a PR onto this branch with these changes https://github.com/DataBiosphere/stairway/pull/111

ddietterich commented on stairway/src/main/java/bio/terra/stairway/impl/FlightFilterAccess.java:
```diff
@@ -150,52 +155,22 @@ String makeSql() {
-    sb.append(" GROUP BY I.flightid) INPUT ON F.flightid = INPUT.flightid")
-        .append(" WHERE INPUT.matchCount = ")
-        .append(filter.getInputPredicates().size());
     makeFlightFilter(sb, " AND ");
```
This means the flight filters cannot participate in the boolean structure. Only added in as a top-level AND.

How about this idea: instead of segregating the generation by maintaining two separate arrays, maybe remember the source of the data in the `FlightFilterPredicate`. We could have the` FlightFilterPredicate` take a `keyType` enum argument. Key type would be FLIGHT_TABLE or INPUT_PARAMETER.

In the future, we could add filtering by the output params of completed flights by adding an OUTPUT_PARAMETER type.

The current `add` method results could be collected into a single list of `List<FlightFilterPredicateInterface.Builder>`.

nmalfroy replied:
I updated the PR so that inputs can be included too.  That was a great call to add.  I've update the unit tests to have examples of those being mixed in too

nmalfroy replied:
I had thought about that but then you get a weird case with supporting the existing API that returns either input or flight filters.  I went a separate approach and left those but added a top level boolean expression that now takes input and flight (and could potentially take output) filters.  So it's really two ways to construct filtering

ddietterich commented on stairway/src/main/java/bio/terra/stairway/FlightFilter.java:
```diff
@@ -233,7 +354,96 @@ public String getParameterName() {
+    private final Operation operation;
+    private final List<FlightFilterPredicateInterface> expressions;
+
+    private FlightBooleanOperationExpression(
```
IntelliJ also thinks this is never used.

nmalfroy replied:
that should be cleaned up now

ddietterich commented on stairway/src/main/java/bio/terra/stairway/impl/FlightFilterAccess.java:
```diff
@@ -70,50 +83,42 @@ void storePredicateValues(NamedParameterPreparedStatement statement)
-   *   ON flight.flightid = flight_input.flightid
-   * WHERE flight-filters
-   *   AND flight_input.key = 'key' AND flight_input.value = 'json-of-object'
+   * SELECT *
```
These examples only cover the cases where the predicates are on input values.
They need to cover the cases where the predicates are on flight columns.

nmalfroy replied:
I'll update the comments here

nmalfroy replied:
I updated the comments to show more examples

ddietterich commented on stairway/src/main/java/bio/terra/stairway/FlightFilter.java:
```diff
@@ -233,7 +354,96 @@ public String getParameterName() {
     public enum Datatype {
       STRING,
       TIMESTAMP,
+      LIST,
```
What this type means isn't obvious to me. And the semantics in FlightFilterAccess that convert to JSON are also not obvious. Maybe some documentation about what this means?

I'm noticing that documentation about using the FlightFilter is only in the comment at the top of this module and is not comprehensive. Not your problem, but something for me to add to the Stairway list.

nmalfroy replied:
I updated the documentation at the top of FlightFilter to address this. I'll add more inline with the code.  Maybe list could be array (which is a little more generic).  I didn't make the LIST type parameterized since I and only imagine it would ever be used with strings (I don't think we need to support List of timestamps for querying)

ddietterich commented on stairway/src/main/java/bio/terra/stairway/impl/FlightFilterAccess.java:
```diff
@@ -289,19 +277,70 @@ private void storeFlightPredicateValue(
+      valuePredicate =
+          "I.value "
+              + predicate.getOp().getSql()
+              + " (SELECT JSON_ARRAY_ELEMENTS_TEXT(CAST(:"
```
This is pretty odd. What is this supposed to do?

nmalfroy replied:
oh, that's in order to work with the JSON encoded data in the flight input table

ddietterich commented on stairway/src/main/java/bio/terra/stairway/FlightFilter.java:
```diff
@@ -233,7 +354,96 @@ public String getParameterName() {
       NULL
     }
+
+    public static class Builder implements FlightFilterPredicateInterface.Builder {
```
Everything in here is public. I wonder if it needs to be.

nmalfroy replied:
I moved away from a builder so this should be a moot point now

ddietterich reviewed (changes requested):
At long last, I got to look at this. I like this approach. Now my concerns are mostly at the implementation level.

My primary concern is the lack of support for flight column predicates in the boolean form. I think there is a nice approach to addressing this, outlined in comments.

I also think the Builder model is not needed and can be accommodated in the FlightFilterAccess code. That would be a nice simplification.

nmalfroy committed:
Add the ability to include flight parameters to the boolean queries

...also made a small tweak to the verification function in EnumerateFlightsTest to make it easier to debug (e.g. you can see what flights were expected and what flights were returned)

pshapiro4broad commented on stairway/src/main/java/bio/terra/stairway/FlightFilter.java:
```diff
@@ -177,12 +236,180 @@ public FlightFilter submittedTimeSortDirection(
 
-  public static class FlightFilterPredicate {
+  public interface FlightFilterPredicateInterface {
+    interface Builder {
```
Note that class members like `Builder` aren't inherited so there's no reason or benefit to have this as a member, other than needing to type `FlightFilterPredicateInterface.Builder` instead of `Builder` when you refer to it.

It's also suspicious to me that `FlightFilterPredicateInterface` is an interface with no members. Looking at the code, I think it's possible to make `storeInputPredicateValue()` and `storeFlightPredicateValue()` both methods on these objects, vs methods in `FlightFilterAccess`. If that works, it should make the code simpler.

nmalfroy replied:
It's really there to enable backwards compatibility with existing filtering methods.  I want this PR to not be a breaking change for existing implementors and this should allow that

pshapiro4broad replied:
I must be missing something, can you provide an example where changing this would break backward compatibility?

nmalfroy replied:
Yes, FlightFilterPredicate is a public interface (not java interface but an object that a user can instantiate). By turning it into an interface, it makes it so that existing code should be able to be left unchanged

nmalfroy replied:
I should amend this: by having `FlightFilterPredicateInterface` this allows nesting or/and clauses by having `FlightBooleanOperationExpression` include a list of `FlightFilterPredicateInterface` (which can be `FlightBooleanOperationExpression` or `FlightFilterPredicate` objects) as it's children

pshapiro4broad committed:
Using the query support changes as a start, move parameter name creation to the SQL code in `FlightFilterAccess` to avoid needing state to create a boolean filter predicate.

pshapiro4broad committed:
Further refactor filter code to avoid needing to type-switch explicitly (in storeInputPredicateValue) and implicitly (due to overriding on the first argument of storeInputPredicateValue()).

nmalfroy committed:
Merge pull request #111 from DataBiosphere/ps/query-support-no-builder

Boolean query support without a builder pattern

nmalfroy committed:
Merge pull request #112 from DataBiosphere/ps/filter-refactor

Further refactor filter code to avoid needing to type-switch

nmalfroy committed:
docstring updates

pshapiro4broad commented on stairway/src/main/java/bio/terra/stairway/FlightFilter.java:
```diff
@@ -23,18 +26,42 @@
+   * builder methods that are created using static methods on this class.
+   */
+  public FlightFilter(FlightBooleanOperationExpression inputBooleanOperationExpression) {
+    this();
```
In general I think it's better to chain from fewer params to more, so it's explicit that `inputBooleanOperationExpression` is `null` in one case. Then you could make `inputBooleanOperationExpression` `final`.

You could also also change the other fields to be initialized at their declarations, but it could work either way.
```
public class FlightFilter {
  private final List<FlightFilterPredicate> flightPredicates = new ArrayList<>();
  private final List<FlightFilterPredicate> inputPredicates = new ArrayList<>();
  private final FlightBooleanOperationExpression inputBooleanOperationExpression;
  private FlightFilterSortDirection submittedTimeSortDirection = FlightFilterSortDirection.ASC;

  // Mapper should be used to deserialize to generic Postgres JSON
  private static final ObjectMapper pgJsonMapper = new ObjectMapper();

  public FlightFilter() {
    this(null);
  }

  /**
   * Use this constructor to allow for more complex selection criteria. This takes in expression
   * builder methods that are created using static methods on this class.
   */
  public FlightFilter(FlightBooleanOperationExpression inputBooleanOperationExpression) {
    this.inputBooleanOperationExpression = inputBooleanOperationExpression;
  }
```

nmalfroy replied:
ah good call!

pshapiro4broad reviewed (approved):
The diff got pretty big but I think it's a little clearer this way. One last comment about constructor chaining, otherwise looks OK to me.

nmalfroy committed:
Change constructor changing and rename variable

ddietterich reviewed (approved):
Thank you Nick and Phil!
